src/device-state/current-state.ts

```diff
@@ -104,170 +109,149 @@ const report = async ({ stateDiff, conf }: StateReport) => {
+	// Calculate differences between current state and last reported state
+	const stateDiff = getStateDiff(lastReportedState, stateForReport);
+	// Try to send report now
+	try {
+		await report(stateDiff);
+	} catch (e) {
+		// Use appUpdatePollInterval as maxDelay since we encountered an error and will backoff now
+		const maxDelay = await config.get('appUpdatePollInterval');
+		// Check if it was an HTTP status error
+		if (e instanceof StatusError) {
+			// Supervisor got an HTTP error from API
+			log.error(
+				`Device state report failure! Status code: ${e.statusCode} - message:`,
+				e.message,
+			);
+			// Use retryAfter as maxDelay if the header is present
+			await backoffPromise(
+				reportHandler,
+				// Do not do exponential backoff if the API reported a retryAfter
+				e.retryAfter ? 0 : attempts,
+				maxDelay,
+				// Start the polling at the value given by the API if any
+				e.retryAfter ?? patchInterval,
+			);
+		} else {
+			// Report did not get a HTTP response
+			eventTracker.track('Device state report failure', {
+				error: e.message,
+			});
+			await backoffPromise(
+				reportHandler,
+				stateReportErrors++,
+				maxDelay,
+				patchInterval,
+			);
+		}
+	}
 
-		// report state diff
-		throttledReport(
-			{ stateDiff, conf },
-			() => {
-				// If the patch succeeds update lastReport and reset the error counter
-				_.assign(lastReportedState.local, stateDiff.local);
-				_.assign(lastReportedState.dependent, stateDiff.dependent);
-				stateReportErrors = 0;
+	// Report was successful so update lastReportedState
+	_.assign(lastReportedState.local, stateDiff.local);
+	_.assign(lastReportedState.dependent, stateDiff.dependent);
+	stateReportErrors = 0;
 
-				reportPending = false;
-			},
-			(e) => {
-				if (e instanceof StatusError) {
-					// We don't want these errors to be classed as a report error, as this will cause
-					// the watchdog to kill the supervisor - and killing the supervisor will
-					// not help in this situation
-					log.error(
-						`Device state report failure! Status code: ${e.statusCode} - message:`,
-						e.message,
-					);
+	// Wait for interval before trying to report again
+	await sleep(patchInterval);
 
-					// We want to backoff on all errors, but without having the healthchecks
-					// get triggered.
-					// This will use retryAfter as maxDelay if the header is present in the
-					// response by the API
-					backoff(
-						reportCurrentState,
-						// Do not do exponential backoff if the API reported a retryAfter
-						e.retryAfter ? 0 : attempts,
-						maxDelay,
-						// Start the polling at the value given by the API if any
-						e.retryAfter ?? MINIMUM_BACKOFF_DELAY,
-					);
-				} else {
-					eventTracker.track('Device state report failure', {
-						error: e.message,
-					});
-					backoff(reportCurrentState, stateReportErrors++, maxDelay);
-				}
-			},
-		);
-	})();
+	// Begin report loop again with 0 attempts
+	return reportHandler(0);
```

Comment:
Do you know if node performs some cleanup of the stack on this type of call? This is technically an infinite recursion over the lifetime of the supervisor, unless node performs some cleanup, eventually you will run out of call stack space right?

Reply:
I think because it's a promise (via async/await) that it'd reset the stack regularly and so it would be ok, but also there's no benefit to returning it so just removing the `return` avoids the issue altogether